internal/datastore/spanner/caveat.go

```diff
@@ -3,23 +3,122 @@ package spanner
 import (
 	"context"
 	"fmt"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/codes"
 
 	"github.com/authzed/spicedb/pkg/datastore"
 	core "github.com/authzed/spicedb/pkg/proto/core/v1"
 )
 
 func (sr spannerReader) ReadCaveatByName(ctx context.Context, name string) (*core.CaveatDefinition, datastore.Revision, error) {
-	return nil, datastore.NoRevision, fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(ctx, "ReadCaveatByName", trace.WithAttributes(attribute.String("name", name)))
+	defer span.End()
+
+	caveatKey := spanner.Key{name}
+	row, err := sr.txSource().ReadRow(ctx, tableCaveat, caveatKey, []string{colCaveatDefinition, colCaveatTS})
+	if err != nil {
+		if spanner.ErrCode(err) == codes.NotFound {
+			return nil, datastore.NoRevision, datastore.NewCaveatNameNotFoundErr(name)
+		}
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadCaveat, err)
+	}
+	var serialized []byte
+	var updated time.Time
+	if err := row.Columns(&serialized, &updated); err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadCaveat, err)
+	}
+
+	loaded := &core.CaveatDefinition{}
+	if err := loaded.UnmarshalVT(serialized); err != nil {
+		return nil, datastore.NoRevision, err
+	}
+	return loaded, revisionFromTimestamp(updated), nil
 }
 
 func (sr spannerReader) ListCaveats(ctx context.Context, caveatNames ...string) ([]*core.CaveatDefinition, error) {
-	return nil, fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(ctx, "ListCaveats", trace.WithAttributes(attribute.StringSlice("names", caveatNames)))
+	defer span.End()
+
+	keyset := spanner.AllKeys()
+	if len(caveatNames) > 0 {
+		keys := make([]spanner.Key, 0, len(caveatNames))
+		for _, n := range caveatNames {
+			keys = append(keys, spanner.Key{n})
+		}
+		keyset = spanner.KeySetFromKeys(keys...)
+	}
+	iter := sr.txSource().Read(
+		ctx,
+		tableCaveat,
+		keyset,
+		[]string{colCaveatDefinition},
+	)
+
+	var caveats []*core.CaveatDefinition
+	if err := iter.Do(func(row *spanner.Row) error {
+		var serialized []byte
+		if err := row.Columns(&serialized); err != nil {
+			return err
+		}
+
+		loaded := &core.CaveatDefinition{}
+		if err := loaded.UnmarshalVT(serialized); err != nil {
+			return err
+		}
+		caveats = append(caveats, loaded)
+
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf(errUnableToListCaveats, err)
+	}
+
+	return caveats, nil
 }
 
 func (rwt spannerReadWriteTXN) WriteCaveats(caveats []*core.CaveatDefinition) error {
-	return fmt.Errorf("unimplemented caveat support in datastore")
+	_, span := tracer.Start(rwt.ctx, "WriteCaveats")
+	defer span.End()
+
+	names := map[string]struct{}{}
```

Comment:
why not relying on the database to impose key uniqueness instead? I can understand it being faster given it does not have to hit the DB, but we still need to rely on the database anyway given an overlapping transaction could attempt to store the same caveat?

Reply:
Not sure why I did this - removed it.

Reply:
Idempotency FTW 😄